spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/ApiKeyAuthenticator.java

```diff
@@ -0,0 +1,150 @@
+ * limitations under the License.
+ * #L%
+ */
+
+import eu.coatrack.api.ApiKey;
+import eu.coatrack.proxy.security.exceptions.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.authentication.AuthenticationManager;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ *  Checks if the API key token value sent by the client is valid. If so, an authentication object
+ *  including the granted authorities is generated, to be further handled by access decision voters.
+ *
+ * @author gr-hovest, Christoph Baier
+ */
+
+@Service
+public class ApiKeyAuthenticator implements AuthenticationManager {
+
+    private static final Logger log = LoggerFactory.getLogger(ApiKeyAuthenticator.class);
+
+    private final LocalApiKeyManager localApiKeyManager;
+    private final ApiKeyFetcher apiKeyFetcher;
+    private final ApiKeyValidator apiKeyValidator;
+    private final Set<SimpleGrantedAuthority> authoritiesGrantedToCoatRackAdminApp = new HashSet<>();
+
+    public ApiKeyAuthenticator(LocalApiKeyManager localApiKeyManager,
+                               ApiKeyFetcher apiKeyFetcher, ApiKeyValidator apiKeyValidator) {
+        this.localApiKeyManager = localApiKeyManager;
+        this.apiKeyFetcher = apiKeyFetcher;
+        this.apiKeyValidator = apiKeyValidator;
+
+        authoritiesGrantedToCoatRackAdminApp.add(new SimpleGrantedAuthority(
+                ServiceApiAccessRightsVoter.ACCESS_SERVICE_AUTHORITY_PREFIX + "refresh"));
+    }
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        if (authentication == null)
+            throw new BadCredentialsException("The authentication was null");
+
+        log.debug("Verifying the authentication {}.", authentication.getName());
+        try {
+            String apiKeyValue = extractApiKeyValueFromAuthentication(authentication);
+            return createAuthTokenIfApiKeyIsValid(apiKeyValue);
+        } catch (BadCredentialsException e) {
+            throw e;
+        } catch (Exception e) {
+            log.warn("The API keys validity could not be decided due to following error: ", e);
+            return null;
+        }
+    }
+
+    private String extractApiKeyValueFromAuthentication(Authentication authentication) {
+        try{
+            log.debug("Getting API key value from authentication {}.", authentication.getName());
+            Assert.notNull(authentication.getCredentials(), "The credentials of " + authentication.getName()
```

Comment:
Assertions are removed at runetime, so you cannot rely on them for exception throwing

Reply:
Please not that these are not "normal Java assertions", but Spring assertions: https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/Assert.html

Reply:
I agree on the comment from @gr-hovest-atb . I tested this to be sure and it does work at runtime.